### DISCUSSION TITLE
TreeView Drag and Drop

### DISCUSSION THREAD

**Discussion by JohnKesko** [AvaloniaUI/Avalonia#13015] [Q&A]
How do I implement Drag and Drop in a TreeView? I tried this but nothing is invoked anywhere. 
Sitting on macOS.

XAML:
```
<!-- Playlist -->
<Grid Grid.Row="2" RowDefinitions="Auto, Auto, *">
    <Label Grid.Row="0" Content="Playlist" FontWeight="Bold" Margin="10 10 0 0" />
    <Separator Grid.Row="1" BorderBrush="DarkGray" />
    <Border Grid.Row="2" DragDrop.AllowDrop="True">
        <TreeView ItemsSource="{Binding PlaylistCollection}"
                  Background="White"
                  SelectedItem="{Binding SelectedPlaylistItem}">
            
            <!-- TreeView Data -->
            <TreeView.ItemTemplate>
                <TreeDataTemplate>
                    <StackPanel Orientation="Horizontal">
                        <materialIcons:MaterialIcon Kind="File" Margin="0 0 5 0" />
                        <TextBlock Text="{Binding Name}" />
                    </StackPanel>
                </TreeDataTemplate>
            </TreeView.ItemTemplate>
        </TreeView>
    </Border>
</Grid>
```

Code behind:
```
public partial class NodeView : UserControl
{
    public NodeView()
    {
        InitializeComponent();
        DataContext = App.ServiceContainer.GetRequiredService<NodeViewViewModel>();
        
        AddHandler(DragDrop.DragOverEvent, DragOver);
        AddHandler(DragDrop.DropEvent, Drop);
        AddHandler(DragDrop.DragLeaveEvent, DragLeave);
    }
    
    private async Task DragOver(object? sender, DragEventArgs e)
    {
        await Dispatcher.UIThread.InvokeAsync(() =>
        {
            Console.WriteLine("DragOver");
        });
    }

    private async Task Drop(object? sender, DragEventArgs e)
    {
        await Dispatcher.UIThread.InvokeAsync(() =>
        {
            Console.WriteLine("Drop");
        });
    }

    private async Task DragLeave(object? sender, DragEventArgs e)
    {
        await Dispatcher.UIThread.InvokeAsync(() =>
        {
            Console.WriteLine("Drag leave");
        });
    }
}
```

**Comment by timunie** [comment]
Do you mean file drag and drop or item drag and drop?

For File DnD you have to add `DraGrop.AllowDrop` iirc

**Reply by JohnKesko** [reply]
Item drag and drop. Just reordering an item inside the tree.

**Reply by timunie** [reply]
look into Avalonia.Behaviors and also check out TreeDataGrid. DragDrop if more for system I/O

**Reply by JohnKesko** [reply]
Yes, got it working with that, thanks, works great!

Btw, I couldn't find any documentation about it, I assume it's on the list of ToDo? :)

**Reply by timunie** [reply]
documentation for what is missing?

**Reply by JohnKesko** [reply]
Where can I read more about drag and drop @ https://v11.docs.avaloniaui.net/ ?

**Reply by timunie** [reply]
the docs are now here: https://docs.avaloniaui.net

BUT probably that part is indeed missing, so please open an issue in the docs repo or add a PR to fix it.

**Reply by ajmcateer** [reply]
@JohnKesko Can you share how you got drag and drop working?

**Reply by JohnKesko** [reply]
In Avalonia.Behaviors there is a sample: https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/tree/master/samples/DragAndDropSample

I modified it for my needs but its a great starting point.

**Reply by MechWarrior99** [reply]
Has there been any update on this? I just want to be able to re order the items in the tree view and chainge parent child relations. Is the only/best option still to use the Behaviors package and write a custom solution using it?
